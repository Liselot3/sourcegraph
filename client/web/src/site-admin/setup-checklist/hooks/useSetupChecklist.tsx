import { add, differenceInDays, formatDistanceStrict } from 'date-fns'

import { gql, useQuery } from '@sourcegraph/http-client'
import { Code, Link, Text } from '@sourcegraph/wildcard'

import { SetupChecklistResult, SetupChecklistVariables } from '../../../graphql-operations'

const QUERY = gql`
    query SetupChecklist {
        site {
            externalServicesCounts {
                remoteExternalServicesCount
            }
            productSubscription {
                license {
                    isValid
                    expiresAt
                    tags
                    userCount
                }
            }
            users(deletedAt: { empty: true }) {
                totalCount
            }
        }
    }
`
export enum SetupChecklistItem {
    AddLicense = 'Add license',
    ConnectCodeHost = 'Connect code host',
}

interface UseSetupChecklistReturnType {
    data: {
        id: SetupChecklistItem
        name: string
        path: string
        info: React.ReactNode
        needsSetup?: string
    }[]
    error?: any
    loading: boolean
}

const NOTIFY_LICENSE_EXPIRATION_DAYS = 7
/**
 * Returns either text of why setup/action is required or undefined if everything is good
 *
 * @param args
 * @param now for testing purposes
 */
export const getLicenseSetupStatus = (
    args?: Pick<SetupChecklistResult['site'], 'users' | 'productSubscription'>,
    now = new Date()
): string | undefined => {
    const license = args?.productSubscription?.license

    if (!license?.isValid) {
        return 'The Sourcegraph license key is invalid.'
    }

    if (license?.tags?.includes('dev')) {
        return
    }

    const expiresAt = license?.expiresAt ? new Date(license.expiresAt) : undefined
    if (expiresAt && differenceInDays(expiresAt, now) <= NOTIFY_LICENSE_EXPIRATION_DAYS) {
        return `The Sourcegraph license ${
            differenceInDays(expiresAt, now) <= 0
                ? 'expired ' + formatDistanceStrict(expiresAt, now) + ' ago. Please, get a new license.' // 'Expired two months ago'
                : 'will expire in ' + formatDistanceStrict(expiresAt, now) + '. Please, renew it soon.' // 'Will expire in two months'
        }`
    }

    if (license?.tags?.includes('plan:free-1')) {
        return 'You are on a free plan. Please, upgrade your license to unlock more features.'
    }

    const userCount = args?.users?.totalCount
    const hasExceededUserCount =
        !!userCount && !!license?.userCount && userCount > license?.userCount && !license?.tags.includes('true-up')
    if (hasExceededUserCount) {
        return 'Your user count has exceeded your license limit. Please, get a new license.'
    }

    return
}

const getCodehostSetupStatus = (
    args?: Pick<SetupChecklistResult['site'], 'externalServicesCounts'>
): string | undefined => {
    const codeHostsCounts = args?.externalServicesCounts?.remoteExternalServicesCount
    // todo: add more checks such as valid code host connections
    if (codeHostsCounts === 0) {
        return 'Add code host connection'
    }
    return
}

export function useSetupChecklist(): UseSetupChecklistReturnType {
    const { data, loading, error } = useQuery<SetupChecklistResult, SetupChecklistVariables>(QUERY, {})

    return {
        data: [
            {
                id: SetupChecklistItem.AddLicense,
                name: 'Add license',
                path: '/site-admin/configuration',
                needsSetup: getLicenseSetupStatus(data?.site),
                info: (
                    <Text className="m-0">
                        Add a new <Code>licenseKey</Code> json field.{' '}
                        <Link to="/help/admin/config/site_config#licenseKey">Learn more</Link>
                    </Text>
                ),
            },
            {
                id: SetupChecklistItem.ConnectCodeHost,
                name: 'Connect codehost(s)',
                path: '/site-admin/external-services/new',
                info: <div>Add codehost connections and sync its repositories</div>,
                needsSetup: getCodehostSetupStatus(data?.site),
            },
        ],
        loading,
        error,
    }
}
