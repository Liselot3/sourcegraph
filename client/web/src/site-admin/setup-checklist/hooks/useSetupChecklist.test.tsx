import { add, endOfYesterday, sub } from 'date-fns'
import { cloneDeep, merge } from 'lodash'

import { getLicenseSetupStatus } from './useSetupChecklist'

type DeepPartial<T> = T extends object
    ? {
          [P in keyof T]?: DeepPartial<T[P]>
      }
    : T

describe('useSetupChecklist', () => {
    describe('getLicenseSetupStatus', () => {
        const now = new Date(2023, 8, 10)
        const args: Parameters<typeof getLicenseSetupStatus>[0] = {
            productSubscription: {
                license: {
                    __typename: 'ProductLicenseInfo',
                    isValid: true,
                    expiresAt: add(now, { days: 100 }).toUTCString(),
                    tags: [],
                    userCount: 10,
                },
            },
            users: {
                totalCount: 9,
            },
        }
        const toArgs = (newArgs: DeepPartial<typeof args>) => merge(cloneDeep(args), newArgs)

        it('returns undefined if license is valid', () => {
            expect(getLicenseSetupStatus(args, now)).toBeUndefined()
        })

        it('returns reason if license is invalid', () => {
            const reason = getLicenseSetupStatus(toArgs({ productSubscription: { license: { isValid: false } } }), now)
            expect(reason).toBeDefined()
            expect(reason).toMatchSnapshot()
        })

        it('returns reason if free plan license', () => {
            const reason = getLicenseSetupStatus(
                toArgs({ productSubscription: { license: { tags: ['plan:free-1'] } } }),
                now
            )
            expect(reason).toBeDefined()
            expect(reason).toMatchSnapshot()
        })

        it('returns reason if exceeded allowed user count', () => {
            const reason = getLicenseSetupStatus(toArgs({ users: { totalCount: 11 } }), now)
            expect(reason).toBeDefined()
            expect(reason).toMatchSnapshot()
        })

        it('returns undefined if exceeded allowed user count but true-up', () => {
            const reason = getLicenseSetupStatus(
                toArgs({ users: { totalCount: 11 }, productSubscription: { license: { tags: ['true-up'] } } }),
                now
            )
            expect(reason).toBeUndefined()
        })

        it('returns reason if expired ', () => {
            const reason = getLicenseSetupStatus(
                toArgs({ productSubscription: { license: { expiresAt: sub(now, { days: 1 }).toUTCString() } } }),
                now
            )
            expect(reason).toBeDefined()
            expect(reason).toMatchSnapshot()
        })

        it('returns reason if expiring soon ', () => {
            const reason = getLicenseSetupStatus(
                toArgs({ productSubscription: { license: { expiresAt: add(now, { days: 6 }).toUTCString() } } }),
                now
            )
            expect(reason).toBeDefined()
            expect(reason).toMatchSnapshot()
        })

        it('returns undefined if dev plan ', () => {
            const reason = getLicenseSetupStatus(
                toArgs({
                    productSubscription: { license: { expiresAt: endOfYesterday().toUTCString(), tags: ['dev'] } },
                }),
                now
            )
            expect(reason).toBeUndefined()
        })
    })
})
