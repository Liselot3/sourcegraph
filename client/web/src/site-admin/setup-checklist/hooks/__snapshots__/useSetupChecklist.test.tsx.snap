// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSetupChecklist getLicenseSetupStatus returns reason if exceeded allowed user count 1`] = `"Your user count has exceeded your license limit. Please, get a new license."`;

exports[`useSetupChecklist getLicenseSetupStatus returns reason if expired  1`] = `"The Sourcegraph license expired 1 day ago. Please, get a new license."`;

exports[`useSetupChecklist getLicenseSetupStatus returns reason if expiring soon  1`] = `"The Sourcegraph license will expire in 6 days. Please, renew it soon."`;

exports[`useSetupChecklist getLicenseSetupStatus returns reason if free plan license 1`] = `"You are on a free plan. Please, upgrade your license to unlock more features."`;

exports[`useSetupChecklist getLicenseSetupStatus returns reason if isValid license 1`] = `"The Sourcegraph license key is invalid."`;
